browser/ui/webui/brave_shields/shields_panel_data_handler.cc

```diff
@@ -184,18 +175,22 @@ void ShieldsPanelDataHandler::OnTabStripModelChanged(
     const TabStripModelChange& change,
     const TabStripSelectionChange& selection) {
   if (selection.active_tab_changed()) {
-    // OnResourcesChanged doesnt get triggered instantly on active tab change so
-    // trigger this explicitly
-    UpdateSiteBlockInfo();
+    // To make logic simpler, always remove observer when active tab changed.
+    // And then, start observing when there is new active web contents.
+    if (shields_data_ctrl_) {
+      shields_data_ctrl_->RemoveObserver(this);
+      shields_data_ctrl_ = nullptr;
+    }
 
     if (selection.new_contents) {
-      BraveShieldsDataController::FromWebContents(selection.new_contents)
-          ->AddObserver(this);
+      shields_data_ctrl_ =
+          BraveShieldsDataController::FromWebContents(selection.new_contents);
+      shields_data_ctrl_->AddObserver(this);
     }
 
-    if (selection.old_contents) {
-      BraveShieldsDataController::FromWebContents(selection.old_contents)
-          ->RemoveObserver(this);
-    }
+    // OnResourcesChanged doesnt get triggered instantly on active tab change so
+    // trigger this explicitly.
+    // Call this after new |shields_data_ctrl_| is set.
+    UpdateSiteBlockInfo();
```

Comment:
shouldn't `UpdateSiteBlockInfo()` called in the if block above?

Reply:
`UpdateSiteBlockInfo()` should be run after `shields_data_ctrl_` is set?

Reply:
Ah, I see. you mean in the `if (selection.new_contents) {`

Reply:
Done.